### PR TITLE
dyno: Fix bugs with deprecation and unstable warnings

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
dyno: Fix bugs with deprecation and unstable warnings

Fix a few bugs that were causing test failures when running tests
with the dyno scope resolver.

Add a query `isTypeDefaultInitializable` that is used to fix a bug
with initializers. The bug was screwing up unit tests.

The query is not complete and is missing some obvious logic like
returning `false` for non-nilable owned classes, or detecting when
a type's default initializer does not exist. I will improve it
in future efforts.

Reviewed by @mppf. Thanks!

TESTING

- [x] `paratest`, `--dyno`

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

### CODE REVIEW

File: frontend/lib/resolution/default-functions.cpp [hunk 1]
```diff
@@ -185,19 +186,27 @@ generateInitSignature(Context* context, const CompositeType* inCompType) {
...
-    bool hasDefault = rf.fieldHasDefaultValue(i);
-    const uast::Decl* node = nullptr;
-
-    auto fd = UntypedFnSignature::FormalDetail(name, hasDefault, node);
+    auto fieldQt = rf.fieldType(i);
+    auto formalName = rf.fieldName(i);
+    bool formalHasDefault = rf.fieldHasDefaultValue(i);
+    const uast::Decl* formalAst = nullptr;
+
+    // A field may not have a default value. If it is default-initializable
```
Comment (mppf) [inline frontend/lib/resolution/default-functions.cpp:194]: Did this fix have anything to do with running the dyno scope resolver? If not, it arguably should be in a different PR.

File: frontend/lib/resolution/resolution-queries.cpp [hunk 2]
```diff
@@ -3326,7 +3326,63 @@ bool isNameOfField(Context* context, UniqueString name, const Type* t) {
...
+
+    // TODO: Do I still need to consider field genericity, here? I.E., if
+    // a field is marked 'GENERIC_WITH_DEFAULTS' is there more to do?
+    // If I can tell the thing is concrete from the ResolvedFields, then
+    // there's probably no need to recurse.
+    if (rf.isGenericWithDefaults()) {
+      for (int i = 0; i < rf.numFields(); i++) {
+        auto ft = rf.fieldType(i).type();
+
+        // Skip to avoid recursive query.
```
Comment (mppf) [inline frontend/lib/resolution/resolution-queries.cpp:3363]: This isn't sufficient because we can have mutually recursive types with classes

**Review by mppf — APPROVED** [review]
Unless there is a reason the default-initialization stuff needs to be in the PR to fix deprecation/unstable in scope resolving, I'd like it to be in a different PR.

File: frontend/lib/resolution/Resolver.cpp [hunk 3]
```diff
@@ -2689,30 +2689,38 @@ getDecoratedClassForNew(Context* context, const New* node,
...
 
-// TODO: Emit warning about 'new borrowed' being unstable.
-// TODO: How do we handle '?'.
-void Resolver::resolveNewForClass(const New* node,
-                                  const ClassType* classType) {
-  ResolvedExpression& re = byPostorder.byAst(node);
-
-  // TODO: Verify initial class type?
-  auto cls = getDecoratedClassForNew(context, node, classType);
+static void resolveNewForClass(Resolver& rv, const New* node,
```
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply frontend/lib/resolution/Resolver.cpp:2692]: I just thought that there was no point in cluttering up the resolver interface. Most of the methods on there are utility methods that can be used broadly, these aren't.

File: frontend/lib/resolution/default-functions.cpp (hunk 1 above)
Comment (dlongnecke-cray) [reply frontend/lib/resolution/default-functions.cpp:194]: Yes. I need to run the full resolver, and init calls failing to resolve were screwing up unit tests.

File: frontend/lib/resolution/resolution-queries.cpp (hunk 2 above)
Comment (dlongnecke-cray) [reply frontend/lib/resolution/resolution-queries.cpp:3363]: I'll have to study how you've handled that later, thanks.
